drivers/photo-electronic/adpd410x/adpd410x.c

```diff
@@ -0,0 +1,660 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include "error.h"
+
+#include "adpd410x.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Read device register.
+ * @param dev - Device handler.
+ * @param address - Register address.
+ * @param data - Pointer to the register value container.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
+			  uint16_t *data)
+{
+	int32_t ret;
+	uint8_t buff[] = {0, 0, 0, 0};
+
+	if(dev->dev_type == ADPD4100) {
+		buff[0] = (address & 0x7f80) >> 7;
+		buff[1] = (address & 0x7f) << 1;
+		buff[1] &= 0xfe;
+
+		ret = spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, 4);
+		if(ret != SUCCESS)
+			return FAILURE;
+	} else {
+		buff[0] = (address & 0x7f00) >> 8;
+		buff[0] |= 0x80;
+		buff[1] = address & 0xff;
+
+		/* No stop bit */
+		ret = i2c_write(dev->dev_ops.i2c_phy_dev, buff, 2, 0);
+		if(ret != SUCCESS)
+			return FAILURE;
+		ret = i2c_read(dev->dev_ops.i2c_phy_dev, (buff + 2), 2, 1);
+		if(ret != SUCCESS)
+			return FAILURE;
+	}
+
+	*data = ((uint16_t)buff[2] << 8) & 0xff00;
+	*data |= buff[3] & 0xff;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write device register.
+ * @param dev - Device handler.
+ * @param address - Register address.
+ * @param data - New register value.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reg_write(struct adpd410x_dev *dev, uint16_t address,
+			   uint16_t data)
+{
+	uint8_t buff[] = {0, 0, 0, 0};
+
+	if(dev->dev_type == ADPD4100) {
+		buff[0] = (address & 0x7f80) >> 7;
+		buff[1] = (address & 0x7f) << 1;
+		buff[1] &= 0xfe;
+		buff[2] = (data & 0xff00) >> 8;
+		buff[3] = data & 0xff;
+
+		return spi_write_and_read(dev->dev_ops.spi_phy_dev, buff, 4);
+	} else {
+		buff[0] = (address & 0x7f00) >> 8;
+		buff[0] |= 0x80;
+		buff[1] = address & 0xff;
+		buff[2] = (data & 0xff00) >> 8;
+		buff[3] = data & 0xff;
+
+		/* No stop bit */
+		return i2c_write(dev->dev_ops.i2c_phy_dev, buff, 4, 1);
+	}
+}
+
+/**
+ * @brief Do a software reset.
+ * @param dev - Device handler.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd410x_reset(struct adpd410x_dev *dev)
```

Comment:
When you do a reset, members in your `dev->` will probably not correspond anymore to the value on the chip and your driver representation of the chip is not accurate. I think that after successful reset you should call a function (something like reset_defaults()) that sets all the parameters of dev-> to default state. There is a catch, some may be non-zero.

Reply:
I had a look at your dev, if seems it doesn't contain any setting that resets to non-zero default value in the chip so this should be ok for now, until such setting is added to the driver.